src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -38,37 +55,241 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
+        /// <param name="log">Logger Instance.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns></returns>
+        private Task<RegistrationPage> GetRegistratioIndexPageAsync(
+            HttpSource httpSource,
+            string rangeUri,
+            string packageId,
+            NuGetVersion lower,
+            NuGetVersion upper,
+            HttpSourceCacheContext httpSourceCacheContext,
+            ILogger log,
+            CancellationToken token)
+        {
+            var packageIdLowerCase = packageId.ToLowerInvariant();
+            var registrationPage = httpSource.GetAsync(
+                            new HttpSourceCachedRequest(
+                                rangeUri,
+                                $"list_{packageIdLowerCase}_range_{lower.ToNormalizedString()}-{upper.ToNormalizedString()}",
+                                httpSourceCacheContext)
+                            {
+                                IgnoreNotFounds = true,
+                            },
+                            httpSourceResult => DeserializeStreamDataAsync<RegistrationPage>(httpSourceResult.Stream, token),
+                            log,
+                            token);
+
+            return registrationPage;
+        }
+
+        /// <summary>
+        /// Process RegistrationPage
+        /// </summary>
+        /// <param name="registrationPage">Nuget registration page.</param>
+        /// <param name="results">Used to return nuget result.</param>
+        /// <param name="range">Nuget version range.</param>
+        /// <param name="includePrerelease">Whether to include PreRelease versions into result.</param>
+        /// <param name="includeUnlisted">Whether to include Unlisted versions into result.</param>
+        private void ProcessRegistrationPage(
+            RegistrationPage registrationPage,
+            List<PackageSearchMetadataRegistration> results,
+            VersionRange range, bool includePrerelease,
+            bool includeUnlisted,
+            MetadataReferenceCache metadataCache)
+        {
+            foreach (RegistrationLeafItem registrationLeaf in registrationPage.Items)
+            {
+                PackageSearchMetadataRegistration catalogEntry = registrationLeaf.CatalogEntry;
+                NuGetVersion version = catalogEntry.Version;
+                bool listed = catalogEntry.IsListed;
+
+                if (range.Satisfies(catalogEntry.Version)
+                    && (includePrerelease || !version.IsPrerelease)
+                    && (includeUnlisted || listed))
+                {
+                    catalogEntry.ReportAbuseUrl = _reportAbuseResource?.GetReportAbuseUrl(catalogEntry.PackageId, catalogEntry.Version);
+                    catalogEntry.PackageDetailsUrl = _packageDetailsUriResource?.GetUri(catalogEntry.PackageId, catalogEntry.Version);
+                    metadataCache.GetObject(catalogEntry);
```

Comment:
The return value of `GetObject` isn't used. Can we stop calling it and get rid of the changes related to reflection?

Reply:
Actually I checked it what this thing is doing, it do have benefit. It depopulate same string types because of `string` is true immutable type. All same content strings point to same memory address, I did manually checked it, it works for string type, so I decided to improve it. It's hard to explain here, but I can talk to you on Teams.

Reply:
I'm not sure how this is helping. This is trying to dedup the full object. I don't see the NuGetVersion and Strings being deduped at all.

Reply:
It only works string property with both getter and setter. I'm not sure what kind of input you used to test it, when I'm debugging it looks working by referencing same content strings to same content. I don't think this one decrease number of NuGetVersions, only decrease overall memory size reserved by string properties of instances passed through this cache.

Reply:
I don't see how the strings are getting deduped. 
I ran Get on the `nuget.packagemanagement` package and the cache was empty at the end.

Reply:
@nkolev92 what do you mean by cache exactly? For example for below case nothing happens, because `a` and `b` are already pointing same object reference:
var a = "aaa";
var b = "aaa";

var res  = Object.ReferenceEquals(a, b);
Console.WriteLine(res); //Expect true
-----------------
But below case you would expect deduplication if you use this cache tool, because they're not same object reference.
var a = "aaa";
var b = "aa";
var c = b+ "a";
var res  = Object.ReferenceEquals(a, c);
Console.WriteLine(res); //expect False

Reply:
I don't know why some code is `bold/large` here.